### PR TITLE
Support gathernd with batch_dim mode

### PR DESCRIPTION
**Pull request by wezuo**
Support gathernd with batch_dim mode.

Example usage scenario: running BERT model with batched input data. When gathering masked lm positions, can used GatherND with batch_dim = 1 for this operation.

Existing framework op that support this: Tensorflow: tf.gather_nd

### CODE REVIEW

File: docs/Changelog.md [hunk 1]
```diff
@@ -14022,6 +13927,130 @@ This version of the operator has been available since version 12 of the default
...
 
+### <a name="GatherND-12"></a>**GatherND-12**</a>
+
+  Given `data` tensor of rank `r` >= 1, `indices` tensor of rank `q` >= 1, and `batch_dims` integer `b`, this operator gathers 
+  slices of `data` into an output tensor of rank `q + r - indices_shape[-1] - 1 - b`.
+  
+  `indices` is an q-dimensional integer tensor, best thought of as a `(q-1)`-dimensional tensor of index-tuples into `data`, 
+  where each element defines a slice of `data`
+  
+  `batch_dims` (denoted as `b`) is an integer indicating the number of bacth dimensions, i.e the first `b` number of dimensions of
```
Comment (hariharans29) [inline docs/Changelog.md]: typo: batch

File: docs/Changelog.md [hunk 2]
```diff
@@ -8184,7 +8184,7 @@ This version of the operator has been available since version 9 of the default O
 
 <dl>
 <dt><tt>input</tt> : T1</dt>
-<dd>1D tensor. The shape of the expected output tensor. If empty tensor is given, the output would be a scalar. All values must be >= 0.</dd>
```
Comment (hariharans29) [inline docs/Changelog.md]: Please fix this. Seems unintentional.

File: docs/Changelog.md [hunk 3]
```diff
@@ -14022,6 +13927,130 @@ This version of the operator has been available since version 12 of the default
...
+### <a name="GatherND-12"></a>**GatherND-12**</a>
+
+  Given `data` tensor of rank `r` >= 1, `indices` tensor of rank `q` >= 1, and `batch_dims` integer `b`, this operator gathers 
+  slices of `data` into an output tensor of rank `q + r - indices_shape[-1] - 1 - b`.
+  
+  `indices` is an q-dimensional integer tensor, best thought of as a `(q-1)`-dimensional tensor of index-tuples into `data`, 
+  where each element defines a slice of `data`
+  
+  `batch_dims` (denoted as `b`) is an integer indicating the number of bacth dimensions, i.e the first `b` number of dimensions of 
+  `data` tensor and `indices` are representing the batches, and the gather starts from the `b`+1 dimension.
```
Comment (hariharans29) [inline docs/Changelog.md]: nit: `b+1` (not `b` + 1), the difference in text highlighted in the documentation might be misleading

File: docs/Changelog.md [hunk 4]
```diff
@@ -14022,6 +13927,130 @@ This version of the operator has been available since version 12 of the default
...
 </dl>
 
+### <a name="GatherND-12"></a>**GatherND-12**</a>
+
+  Given `data` tensor of rank `r` >= 1, `indices` tensor of rank `q` >= 1, and `batch_dims` integer `b`, this operator gathers 
+  slices of `data` into an output tensor of rank `q + r - indices_shape[-1] - 1 - b`.
+  
+  `indices` is an q-dimensional integer tensor, best thought of as a `(q-1)`-dimensional tensor of index-tuples into `data`, 
+  where each element defines a slice of `data`
+
```
Comment (hariharans29) [inline docs/Changelog.md]: should there be a statement saying the first `b` dims of both the `data` and `indices` tensor must be equal ?

File: docs/Changelog.md [hunk 5]
```diff
@@ -14022,6 +13927,130 @@ This version of the operator has been available since version 12 of the default
 <dd>Constrain input and output types to all numerical tensor types.</dd>
 </dl>
 
+### <a name="GatherND-12"></a>**GatherND-12**</a>
+
+  Given `data` tensor of rank `r` >= 1, `indices` tensor of rank `q` >= 1, and `batch_dims` integer `b`, this operator gathers 
+  slices of `data` into an output tensor of rank `q + r - indices_shape[-1] - 1 - b`.
```
Comment (hariharans29) [inline docs/Changelog.md]: The output rank computation here doesn't align with the new example added...

File: docs/Changelog.md (hunk 1 above)
Comment (hariharans29) [inline docs/Changelog.md]: nit: does first `b` -> leading `b` read better ?

File: onnx/defs/tensor/defs.cc [hunk 6]
```diff
@@ -2469,22 +2474,24 @@ The output is computed as follows:
...
  
-1) If `indices_shape[-1] > r` => error condition
+1) If `indices_shape[-1] > r-b` => error condition
 
-2) If `indices_shape[-1] == r`, since the rank of `indices` is `q`, `indices` can be thought of as a `(q-1)`-dimensional tensor
-   containing 1-D tensors of dimension `r`. Let us think of each such `r` ranked tensor as `indices_slice`. 
-   Each *scalar value* corresponding to `data[indices_slice]` is filled into the corresponding location of the `(q-1)`-dimensional tensor 
-   to form the `output` tensor (Example 1 below)
+2) If `indices_shape[-1] == r-b`, since the rank of `indices` is `q`, `indices` can be thought of as `N` `(q-b-1)`-dimensional tensors
+   containing 1-D tensors of dimension `r-b`. Where `N` is an integer equals to the production of 1 and all the elements in the batch dimensions
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc]: Change ". Where" to ", where"

Comment (gramalingam) [reply onnx/defs/tensor/defs.cc]: Change "an integer equals to the production of 1 and" to "the product of"

File: onnx/defs/tensor/defs.cc [hunk 7]
```diff
@@ -2555,6 +2584,8 @@ ONNX_OPERATOR_SET_SCHEMA(
               ctx.getInputType(1)->tensor_type().shape();
           const auto indices_rank = indices_shape.dim_size();
 
+          auto batch_dims = ctx.getAttribute("batch_dims");
+          int64_t batch_dims_data = batch_dims ? static_cast<int>(batch_dims->i()) : 0;
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc]: The above two lines can be simplified to:
```
   int64_t batch_dims_data = getAttribute(ctx, "batch_dims", 0);
```

File: docs/Changelog.md [hunk 8]
```diff
@@ -14022,6 +13927,130 @@ This version of the operator has been available since version 12 of the default
...
+     It is an error if any of the index values are out of bounds.
+  
+  The output is computed as follows:
+  
+  The output tensor is obtained by mapping each index-tuple in the `indices` tensor to the corresponding slice of the input `data`.
+   
+  1) If `indices_shape[-1] > r-b` => error condition
+  
+  2) If `indices_shape[-1] == r-b`, since the rank of `indices` is `q`, `indices` can be thought of as `N` `(q-b-1)`-dimensional tensors
+     containing 1-D tensors of dimension `r-b`. Where `N` is an integer equals to the production of 1 and all the elements in the batch dimensions
```
Comment (hariharans29) [inline docs/Changelog.md]: production -> product

File: docs/Changelog.md [hunk 9]
```diff
@@ -14165,6 +14165,134 @@ This version of the operator has been available since version 12 of the default
...
+  `batch_dims` (denoted as `b`) is an integer indicating the number of batch dimensions, i.e the leading `b` number of dimensions of 
+  `data` tensor and `indices` are representing the batches, and the gather starts from the `b+1` dimension. 
+  
+  Some salient points about the inputs' rank and shape:
+   
+  1) r >= 1 and q >= 1 are to be honored. There is no dependency condition to be met between ranks `r` and `q`
+  
+  2) The first `b` dimensions of the shape of `indices` tensor and `data` tensor must be equal.
+  
+  3) b < min(q, r) is to be hornored.
```
Comment (hariharans29) [inline docs/Changelog.md]: nit: honored

File: docs/Changelog.md [hunk 10]
```diff
@@ -14165,6 +14165,134 @@ This version of the operator has been available since version 12 of the default
...
+  
+  3) The `indices_shape[-1]` should have a value between 1 (inclusive) and rank `r-b` (inclusive)
```
Comment (hariharans29) [inline docs/Changelog.md]: the numbering order is incorrect...

File: onnx/defs/tensor/defs.cc [hunk 11]
```diff
@@ -2557,6 +2588,7 @@ ONNX_OPERATOR_SET_SCHEMA(
               ctx.getInputType(1)->tensor_type().shape();
           const auto indices_rank = indices_shape.dim_size();
 
+		  int64_t batch_dims_data = getAttribute(ctx, "batch_dims", 0);
```
Comment (lara-hdr) [inline onnx/defs/tensor/defs.cc]: nit: tabs -> spaces

**Comment by spandantiwari** [comment]
@wezuo - Could you please add some details in the description as to what scenario is this addition needed to support. Is there any existing framework op that will be supported by this?

**Comment by wezuo** [comment]
@spandantiwari , could you point me the appropriate place (e.g., which file) to add those note?

**Comment by spandantiwari** [comment]
> @spandantiwari , could you point me the appropriate place (e.g., which file) to add those note?

@wezuo  - just right at the top of this PR in the opening comment.

**Pull request merged**